File: packages/og-injector/src/ServerApi.ts
import axios, { AxiosRequestConfig, Method } from 'axios'
import { SERVER_URL } from '../config'
import { PostData } from './models/PostData'
import { logger } from './Logger'
import { TagData } from './models/TagData'

const apiClient = axios.create({
  baseURL: SERVER_URL,
})

function request<Res>(method: Method, path: string, config: AxiosRequestConfig = {}): Promise<Res> {
  logger.info(`ServerAPI: ${method} ${SERVER_URL}${path}`)

  return apiClient.request<Res>({ method, url: path, ...config })
    .then(res => res.data)
}

export function getPost(postId: string): Promise<PostData> {
  type Response = { post: PostData }

  return request<Response>('GET', `/post/${postId}`)
    .then(res => res.post)
}

export function getTag(tagId: string): Promise<TagData> {
  type Response = { tag: TagData }

  return request<Response>('GET', `/tag/${tagId}`)
    .then(res => res.tag)
}

export function getTaggedPosts(tagId: string): Promise<PostData[]> {
  type Response = { posts: PostData[] }

  return request<Response>('GET', `/tag/${tagId}/posts`)
    .then(res => res.posts)
}

export function getFirstPost(): Promise<PostData> {
  type Response = { post: PostData }

  return request<Response>('GET', `/post/first`)
    .then(res => res.post)
}

export function getLatestPost(): Promise<PostData> {
  type Response = { post: PostData }

  return request<Response>('GET', `/post/latest`)
    .then(res => res.post)
}

File: packages/og-injector-rust/src/server_api/server_api.rs
use reqwest::{Client, ClientBuilder};
use url::Url;

use crate::config::FeatureFlag::Enabled;
use crate::error::AppResult;
use crate::server_api::models::PostData;
use crate::server_api::models::TagData;
use crate::server_api::responses::{GetPostRes, GetTaggedPostsRes, GetTagRes};

use super::ServerConfig;

pub struct ServerApi {
    api_url: Url,
    client: Client,
}

impl ServerApi {
    pub async fn new(config: ServerConfig) -> AppResult<ServerApi> {
        let mut client = ClientBuilder::new();

        if matches!(config.https, Enabled) && config.ssl_crt.is_some() {
            let cert = config.load_cert().await?;
            if cert.is_some() {
                client = client.add_root_certificate(cert.unwrap());
            }
        }

        let client = client.build()?;
        let api_url = config.url.to_owned();
        Ok(ServerApi { client, api_url })
    }

    fn build_url(&self, path: &str) -> String {
        let mut url = self.api_url.clone();
        url.set_path(path);
        url.to_string()
    }

    pub async fn get_post(&self, post_id: &str) -> AppResult<PostData> {
        let path = format!("/post/{post_id}");
        let url = self.build_url(&path);

        let res = self.client.get(url).send().await?;
        let body = res.text().await?;

        let res: GetPostRes = serde_json::from_str(&body)?;
        Ok(res.post)
    }

    pub async fn get_latest(&self) -> AppResult<PostData> {
        let url = self.build_url("/post/latest");

        let res = self.client.get(url).send().await?;
        let body = res.text().await?;

        let res: GetPostRes = serde_json::from_str(&body)?;
        Ok(res.post)
    }

    pub async fn get_tag(&self, tag_id: &str) -> AppResult<TagData> {
        let path = format!("/tag/{tag_id}");
        let url = self.build_url(&path);

        let res = self.client.get(url).send().await?;
        let body = res.text().await?;

        let res: GetTagRes = serde_json::from_str(&body)?;
        Ok(res.tag)
    }

    pub async fn get_tagged_posts(&self, tag_id: &str) -> AppResult<Vec<PostData>> {
        let path = format!("/tag/{tag_id}/posts");
        let url = self.build_url(&path);

        let res = self.client.get(url).send().await?;
        let body = res.text().await?;

        let res: GetTaggedPostsRes = serde_json::from_str(&body)?;
        Ok(res.posts)
    }
}
